File: PyScripts/blockchain/generate_blockchain.py
import hashlib
import time

class Block:
    def __init__(self, index, timestamp, data, previous_hash):
        self.index = index
        self.timestamp = timestamp
        self.data = data
        self.previous_hash = previous_hash
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        return hashlib.sha256(
            str(self.index).encode('utf-8') +
            str(self.timestamp).encode('utf-8') +
            str(self.data).encode('utf-8') +
            str(self.previous_hash).encode('utf-8')
        ).hexdigest()

class Blockchain:
    def __init__(self):
        self.chain = [self.create_genesis_block()]

    def create_genesis_block(self):
        # Create the first block (genesis block)
        return Block(0, time.time(), "Genesis Block", "0")

    def get_latest_block(self):
        # Return the latest block in the chain
        return self.chain[-1]

    def add_block(self, new_block):
        # Add a new block to the chain
        new_block.previous_hash = self.get_latest_block().hash
        new_block.hash = new_block.calculate_hash()
        self.chain.append(new_block)

# Create a blockchain instance
my_blockchain = Blockchain()

# Add blocks to the blockchain
my_blockchain.add_block(Block(1, time.time(), "Transaction Data", ""))
my_blockchain.add_block(Block(2, time.time(), "More Transactions", ""))

File: Typescript/blockchain/generateBlockchain.ts
import * as crypto from 'crypto';

class Block {
    public index: number;
    public timestamp: number;
    public data: any;
    public previousHash: string;
    public hash: string;
    public nonce: number;

    constructor(index: number, timestamp: number, data: any, previousHash: string = '') {
        this.index = index;
        this.timestamp = timestamp;
        this.data = data;
        this.previousHash = previousHash;
        this.hash = this.calculateHash();
        this.nonce = 0;
    }

    calculateHash(): string {
        return crypto.createHash('sha256').update(this.index + this.timestamp + JSON.stringify(this.data) + this.previousHash + this.nonce).digest('hex');
    }

    mineBlock(difficulty: number): void {
        while (this.hash.substring(0, difficulty) !== Array(difficulty + 1).join('0')) {
            this.nonce++;
            this.hash = this.calculateHash();
        }
        console.log(`Block mined: ${this.hash}`);
    }
}

class Blockchain {
    public chain: Block[];
    public difficulty: number;

    constructor() {
        this.chain = [this.createGenesisBlock()];
        this.difficulty = 4; // Adjust difficulty as needed for mining
    }

    createGenesisBlock(): Block {
        return new Block(0, Date.now(), 'Genesis Block', '0');
    }

    getLatestBlock(): Block {
        return this.chain[this.chain.length - 1];
    }

    addBlock(newBlock: Block): void {
        newBlock.previousHash = this.getLatestBlock().hash;
        newBlock.mineBlock(this.difficulty);
        this.chain.push(newBlock);
    }

    isValidChain(): boolean {
        for (let i = 1; i < this.chain.length; i++) {
            const currentBlock = this.chain[i];
            const previousBlock = this.chain[i - 1];

            if (currentBlock.hash !== currentBlock.calculateHash() || currentBlock.previousHash !== previousBlock.hash) {
                return false;
            }
        }
        return true;
    }
}

// Usage example:
const myBlockchain = new Blockchain();
console.log('Mining block 1...');
myBlockchain.addBlock(new Block(1, Date.now(), { amount: 4 }));
console.log('Mining block 2...');
myBlockchain.addBlock(new Block(2, Date.now(), { amount: 8 }));

console.log('Blockchain is valid:', myBlockchain.isValidChain());
console.log(JSON.stringify(myBlockchain, null, 2));
